File: myExpenses/src/main/java/org/totschnig/myexpenses/dialog/BaseDialogFragment.java
//on some occasions, upon showing a DialogFragment we run into
//"java.lang.IllegalStateException: Can not perform this action after onSaveInstanceState"
//we catch this here, and ignore silently, which hopefully should be save, since activity is being paused
//https://code.google.com/p/android/issues/detail?id=23096#c4

package org.totschnig.myexpenses.dialog;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.snackbar.Snackbar;

import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.preference.PrefHandler;
import org.totschnig.myexpenses.ui.SnackbarAction;
import org.totschnig.myexpenses.util.UiUtils;
import org.totschnig.myexpenses.util.crashreporting.CrashHandler;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentActivity;
import timber.log.Timber;

public abstract class BaseDialogFragment extends DialogFragment {

  protected View dialogView;
  protected LayoutInflater layoutInflater;
  private Snackbar snackbar;

  @Inject
  PrefHandler prefHandler;

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    ((MyApplication) requireActivity().getApplication()).getAppComponent().inject(this);
  }

  protected AlertDialog.Builder initBuilder() {
    MaterialAlertDialogBuilder materialAlertDialogBuilder = new MaterialAlertDialogBuilder(requireContext());
    layoutInflater = LayoutInflater.from(materialAlertDialogBuilder.getContext());
    return materialAlertDialogBuilder;
  }

  protected AlertDialog.Builder initBuilderWithView(int layoutResourceId) {
    AlertDialog.Builder builder = initBuilder();
    //noinspection InflateParams
    dialogView = layoutInflater.inflate(layoutResourceId, null);
    builder.setView(dialogView);
    return builder;
  }

  public void report(IllegalStateException e) {
    @Nullable final FragmentActivity activity = getActivity();
    if (activity == null) {
      Timber.w("Activity is null");
    } else {
      Timber.w("Activity is finishing?: %b", activity.isFinishing());
    }
    CrashHandler.report(e);
  }

  @Override
  public void dismiss() {
    try {
      super.dismiss();
    } catch (IllegalStateException e) {
      report(e);
    }
  }

  protected void showSnackbar(int resId) {
    showSnackbar(getString(resId), Snackbar.LENGTH_LONG, null);
  }

  public void showSnackbar(CharSequence message, int duration, SnackbarAction snackbarAction) {
    View view = dialogView != null ? dialogView : getDialog().getWindow().getDecorView();
    snackbar = Snackbar.make(view, message, duration);
    UiUtils.increaseSnackbarMaxLines(snackbar);
    if (snackbarAction != null) {
      snackbar.setAction(snackbarAction.resId, snackbarAction.listener);
    }
    snackbar.show();
  }

  protected void dismissSnackbar() {
    if (snackbar != null) {
      snackbar.dismiss();
    }
  }

}

File: myExpenses/src/main/java/org/totschnig/myexpenses/dialog/BaseDialogFragment.kt
package org.totschnig.myexpenses.dialog

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TableLayout
import androidx.appcompat.app.AlertDialog
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.lifecycleScope
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.Snackbar
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.preference.PrefHandler
import org.totschnig.myexpenses.ui.SnackbarAction
import org.totschnig.myexpenses.util.crashreporting.CrashHandler
import org.totschnig.myexpenses.util.linkInputsWithLabels
import org.totschnig.myexpenses.util.ui.UiUtils
import timber.log.Timber
import javax.inject.Inject


abstract class BaseDialogFragment : DialogFragment() {
    protected lateinit var dialogView: View
    protected lateinit var materialLayoutInflater: LayoutInflater

    private var snackBar: Snackbar? = null

    @Inject
    lateinit var prefHandler: PrefHandler

    @Inject
    lateinit var dataStore: DataStore<Preferences>


    open val fullScreenIfNotLarge = false

    protected open fun configureArguments(requestKey: String) = Bundle(1).apply {
        putString(KEY_REQUEST_KEY, requestKey)
    }

    val requestKey: String
        get() = requireArguments().getString(KEY_REQUEST_KEY)!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        injector.inject(this)
    }

    private val isLarge
        get() = resources.getBoolean(R.bool.isLarge)

    protected val fullScreen
        get() = fullScreenIfNotLarge && !isLarge

    override fun onStart() {
        super.onStart()
        if (fullScreen) {
            dialog?.window?.let {
                val width = ViewGroup.LayoutParams.MATCH_PARENT
                val height = ViewGroup.LayoutParams.MATCH_PARENT
                it.setLayout(width, height)
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ViewCompat.setOnApplyWindowInsetsListener(view) { v, insets ->
            val systemBarsAndCutouts = insets.getInsets(
                WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()
            )

            v.setPadding(
                systemBarsAndCutouts.left,
                systemBarsAndCutouts.top,
                systemBarsAndCutouts.right,
                systemBarsAndCutouts.bottom
            )

            insets
        }
    }

    @SuppressLint("UseGetLayoutInflater")
    open fun initBuilder(): AlertDialog.Builder = (if (fullScreen)
        AlertDialog.Builder(requireContext(), R.style.FullscreenDialog)
    else MaterialAlertDialogBuilder(requireContext())
            ).also {
            materialLayoutInflater = LayoutInflater.from(it.context)
        }

    protected fun initBuilderWithLayoutResource(layoutResourceId: Int) =
        initBuilderWithView {
            it.inflate(layoutResourceId, null)
        }

    protected fun initBuilderWithView(inflate: (LayoutInflater) -> View) =
        initBuilder().also { builder ->
            dialogView = inflate(materialLayoutInflater).also { view ->
                view.findViewById<TableLayout>(R.id.FormTable)?.let {
                    linkInputsWithLabels(it)
                }
            }
            builder.setView(dialogView)
        }

    fun report(e: IllegalStateException) {
        activity?.also {
            Timber.w("Activity is finishing?: %b", it.isFinishing)
        } ?: run {
            Timber.w("Activity is null")
        }
        CrashHandler.report(e)
    }

    override fun dismiss() {
        try {
            super.dismiss()
        } catch (e: IllegalStateException) {
            report(e)
        }
    }

    protected fun showSnackBar(resId: Int, duration: Int = Snackbar.LENGTH_LONG) {
        showSnackBar(getString(resId), duration)
    }

    fun showSnackBar(
        message: CharSequence,
        duration: Int = Snackbar.LENGTH_LONG,
        snackBarAction: SnackbarAction? = null,
    ) {
        snackBar = Snackbar.make(snackBarContainer, message, duration).also {
            UiUtils.increaseSnackbarMaxLines(it)
            if (snackBarAction != null) {
                it.setAction(snackBarAction.label, snackBarAction.listener)
            }
            it.show()
        }
    }

    fun showDetails(transactionId: Long) {
        lifecycleScope.launchWhenResumed {
            TransactionDetailFragment.show(
                transactionId,
                parentFragmentManager
            )
        }
    }

    protected val snackBarContainer
        get() = if (::dialogView.isInitialized) dialogView else {
            CrashHandler.report(Exception("lateinit property dialogView has not been initialized"))
            dialog!!.window!!.decorView
        }

    protected fun dismissSnackBar() {
        snackBar?.dismiss()
    }
}